File: doc/codes/python/examples/tutorial/silencer.py
import asyncio

import numpy as np

import autd3_link_ethercrab as ethercrab
from autd3 import Client, ClientConfig
from autd3.commands import FixedCompletionTime, FociStm, FociStmOption, SetSilencer
from autd3.geometry import Autd3, Geometry
from autd3.units import Hz
from autd3.value import ControlPoint, ControlPoints, Intensity
from autd3_core import Duration

# xtask:expect-error


async def main() -> None:
    geometry = Geometry([Autd3([0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0])])

    client = await Client.open(
        geometry,
        ethercrab.EtherCrabLinkOption(),
        ClientConfig(),
    )

    center = geometry.center() + np.array([0.0, 0.0, 150.0])
    radius = 30.0

    # ANCHOR: disable
    foci = []
    for i in range(20):
        theta = 2.0 * np.pi * i / 20.0
        p = center + np.array([radius * np.cos(theta), radius * np.sin(theta), 0.0])
        foci.append(ControlPoints([ControlPoint(p)], Intensity.MAX))
    builder = client.datagram_builder()
    builder.push(SetSilencer.disable())
    builder.push(
        FociStm(
            50.0 * Hz,
            foci,
            FociStmOption(),
        )
    )
    for frame in builder.build():
        await client.send_checked(frame)
    # ANCHOR_END: disable

    # ANCHOR: err
    foci = []
    for i in range(40):
        theta = 2.0 * np.pi * i / 40.0
        p = center + np.array([radius * np.cos(theta), radius * np.sin(theta), 0.0])
        foci.append(ControlPoints([ControlPoint(p)], Intensity.MAX))
    builder = client.datagram_builder()
    builder.push(SetSilencer())
    builder.push(
        FociStm(
            50.0 * Hz,
            foci,
            FociStmOption(),
        )
    )
    for frame in builder.build():
        await client.send_checked(frame)
    # ANCHOR_END: err

    # ANCHOR: workaround
    foci = []
    for i in range(40):
        theta = 2.0 * np.pi * i / 40.0
        p = center + np.array([radius * np.cos(theta), radius * np.sin(theta), 0.0])
        foci.append(ControlPoints([ControlPoint(p)], Intensity.MAX))
    builder = client.datagram_builder()
    builder.push(
        SetSilencer(
            FixedCompletionTime(
                intensity=Duration.from_micros(500),
                phase=Duration.from_micros(500),
                strict_mode=True,
            )
        )
    )
    builder.push(FociStm(50.0 * Hz, foci, FociStmOption()))
    for frame in builder.build():
        await client.send_checked(frame)
    # ANCHOR_END: workaround

    await client.close()


if __name__ == "__main__":
    asyncio.run(main())

File: doc/codes/rust/examples/tutorial/silencer.rs
use std::f32::consts::PI;
use std::time::Duration;

use anyhow::Result;

use autd3_rs::commands::{FixedCompletionTime, FociStm, FociStmOption, SetSilencer};
use autd3_rs::geometry::{Autd3, Geometry, Vector3, offset};
use autd3_rs::units::{Hz, mm};
use autd3_rs::value::{ControlPoint, ControlPoints, Intensity};
use autd3_rs::{Client, ClientConfig};
use autd3_rs_link_nop::Nop;

#[tokio::main(flavor = "multi_thread")]
async fn main() -> Result<()> {
    let geometry = Geometry::new(vec![Autd3::default()]);

    let client = Client::open(&geometry, Nop, ClientConfig::default()).await?;

    let center = geometry.center() + offset(0.0 * mm, 0.0 * mm, 150.0 * mm);
    let radius = (30.0 * mm).mm();

    // ANCHOR: disable
    let foci: Vec<ControlPoints<1>> = (0..20)
        .map(|i| {
            let theta = 2.0 * PI * i as f32 / 20.0;
            let p = center + Vector3::new(radius * theta.cos(), radius * theta.sin(), 0.0);
            ControlPoints::new([ControlPoint::from(p)], Intensity::MAX)
        })
        .collect();
    let mut builder = client.datagram_builder();
    builder.push(SetSilencer::disable()).push(FociStm::new(
        50.0 * Hz,
        &foci,
        FociStmOption::default(),
    ));
    for frame in &builder.build()? {
        client.send_checked(frame).await?;
    }
    // ANCHOR_END: disable

    // ANCHOR: err
    let foci: Vec<ControlPoints<1>> = (0..40)
        .map(|i| {
            let theta = 2.0 * PI * i as f32 / 40.0;
            let p = center + Vector3::new(radius * theta.cos(), radius * theta.sin(), 0.0);
            ControlPoints::new([ControlPoint::from(p)], Intensity::MAX)
        })
        .collect();
    let mut builder = client.datagram_builder();
    builder.push(SetSilencer::default()).push(FociStm::new(
        50.0 * Hz,
        &foci,
        FociStmOption::default(),
    ));
    for frame in &builder.build()? {
        client.send_checked(frame).await?;
    }
    // ANCHOR_END: err

    // ANCHOR: workaround
    let foci: Vec<ControlPoints<1>> = (0..40)
        .map(|i| {
            let theta = 2.0 * PI * i as f32 / 40.0;
            let p = center + Vector3::new(radius * theta.cos(), radius * theta.sin(), 0.0);
            ControlPoints::new([ControlPoint::from(p)], Intensity::MAX)
        })
        .collect();
    let mut builder = client.datagram_builder();
    builder
        .push(SetSilencer::new(FixedCompletionTime {
            intensity: Duration::from_micros(500),
            phase: Duration::from_micros(500),
            strict_mode: true,
        }))
        .push(FociStm::new(50.0 * Hz, &foci, FociStmOption::default()));
    for frame in &builder.build()? {
        client.send_checked(frame).await?;
    }
    // ANCHOR_END: workaround

    client.close().await?;
    Ok(())
}
